src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs

```diff
@@ -100,9 +100,7 @@ public Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library, NuGe
             // Merge in runtime dependencies
             if (dependencies?.Count > 0)
             {
-                var nodeDependencies = node.Item.Data.Dependencies.AsList();
-
-                foreach (var nodeDep in nodeDependencies)
+                foreach (var nodeDep in node.Item.Data.Dependencies)
```

Comment:
This still doesn't avoid IEnumerable how about keep it same as old and just change from `foreach` to `for` loop?

Reply:
This isn't intended to avoid IEnumerable. It avoids creating a new list every time which is more expensive than the enumerator I believe.

This could be improved further by checking if this equal to Enumerable.Empty by reference, if it isn't then checking if can be cast to an array. For that I would want more perf testing done on this.

Reply:
`This isn't intended to avoid IEnumerable. It avoids creating a new list every time which is more expensive than the enumerator I believe.`

I'm not sure which is more expensive. What are the improvements after this change? And I like your idea about checking reference type, I think getting rid of IEnumerable may give better results here.

Reply:
Alright, I can profile this specifically. I think those types of changes need to be made everywhere in the method. The intent of this PR was scoped to just this line which I consider a bug.

Reply:
My only point is, if we're updating something then we should update it correctly. Of course current code has a bug since it does double enumeration.

Reply:
Incremental vs a large PR with all the fixes.